src/apdu/dispatcher.c

```diff
@@ -63,23 +54,31 @@ int apdu_dispatcher(const command_t *cmd) {
             buf.size = cmd->lc;
             buf.offset = 0;
 
-            return handler_get_public_key(&buf, (bool) cmd->p1);
+            return handler_get_xpub(&buf);
         case SIGN_TX:
-            if ((cmd->p1 == P1_START && cmd->p2 != P2_MORE) ||  //
-                cmd->p1 > P1_MAX ||                             //
-                (cmd->p2 != P2_LAST && cmd->p2 != P2_MORE)) {
+            if (cmd->p1 > SIGN_TX_P1_MAX ||                                     //
+                (cmd->p1 == SIGN_TX_P1_SIGN && cmd->p2 != SIGN_TX_P2_SIGN) ||   //
+                (cmd->p1 == SIGN_TX_P1_DONE && cmd->p2 != SIGN_TX_P2_DONE)) {   //
                 return io_send_sw(SW_WRONG_P1P2);
             }
+            // if ((cmd->p1 == P1_START && cmd->p2 != P2_MORE) ||  //
+            //     cmd->p1 > P1_MAX ||                             //
+            //     (cmd->p2 != P2_LAST && cmd->p2 != P2_MORE)) {
+            //     return io_send_sw(SW_WRONG_P1P2);
+            // }
 
-            if (!cmd->data) {
-                return io_send_sw(SW_WRONG_DATA_LENGTH);
-            }
+            // XXX: maybe validate data_len for each command?
```

Comment:
I think we must do it. Buffer overflow can be exploited to RCE, which might be dangerous.